internal/cri/server/stats_collector.go

```diff
@@ -0,0 +1,300 @@
+//go:build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/containerd/cgroups/v3"
+	"github.com/containerd/cgroups/v3/cgroup1"
+	cg1 "github.com/containerd/cgroups/v3/cgroup1/stats"
+	cgroupsv2 "github.com/containerd/cgroups/v3/cgroup2"
+	cg2 "github.com/containerd/cgroups/v3/cgroup2/stats"
+	"github.com/containerd/log"
+	"github.com/containerd/typeurl/v2"
+
+	"github.com/containerd/containerd/api/services/tasks/v1"
+	"github.com/containerd/containerd/v2/internal/cri/store/stats"
+	ctrdutil "github.com/containerd/containerd/v2/internal/cri/util"
+)
+
+const (
+	// defaultCollectionInterval is how often the collector fetches metrics.
+	// This matches cAdvisor's default housekeeping interval.
+	defaultCollectionInterval = 1 * time.Second
```

Comment:
maybe these should be tunable values, 120 1sec intervals / each container seems arbitrary and expensive

Reply:
2 mins comes from here - https://github.com/google/cadvisor/blob/b21c1a151105736b8307e8ae88bde5e9a278ac7f/cmd/storagedriver.go#L38

1 second comes from here - https://github.com/google/cadvisor/blob/b21c1a151105736b8307e8ae88bde5e9a278ac7f/manager/container.go#L50

(trying to match what we have currently!)

Reply:
whoops! on it